README.md


# Stack Overflow in Portuguese Tab

<p align="center">
<a href="https://packagist.org/packages/mateusjunges/ignition-stackoverflow-portuguese" target="_blank"><img src="https://poser.pugx.org/mateusjunges/ignition-stackoverflow-portuguese/d/total.svg" alt="Total Downloads"></a>
<a href="https://packagist.org/packages/mateusjunges/ignition-stackoverflow-portuguese" target="_blank"><img src="https://poser.pugx.org/mateusjunges/ignition-stackoverflow-portuguese/v/stable.svg" alt="Latest Stable Version"></a>
<a href="https://packagist.org/packages/mateusjunges/ignition-stackoverflow-portuguese" target="_blank"><img src="https://poser.pugx.org/mateusjunges/ignition-stackoverflow-portuguese/license.svg" alt="License"></a>
<a href="https://github.styleci.io/repos/207148235" target="_blank"><img src="https://github.styleci.io/repos/207148235/shield?style=flat"></a>
<a href="https://travis-ci.org/mateusjunges/ignition-stackoverflow-portuguese"><img src="https://img.shields.io/travis/mateusjunges/ignition-stackoverflow-portuguese/master.svg?style=flat" alt="Build Status"></a>
</p>

This package adds a tab to your ignition screen, which shows matching StackOverflow in Portuguese questions, including a search field.

![StackOverflow PT-BR Tab](stackoverflow-pt-br-tab.png)

## Installation

You can install the package in to a Laravel app that uses [Ignition](https://flareapp.io) via composer:

```bash
composer require mateusjunges/ignition-stackoverflow-portuguese
```

## Usage

Click on the "StackOverflow PT-BR" tab on your Ignition screen to see the tool provided by this package.

### Testing

``` bash
composer test
```

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

### Security

If you discover any security related issues, please email contato@mateusjunges.com instead of using the issue tracker.

## Credits

- [Mateus Junges](https://github.com/mateusjunges)
- [Gummibeer](https://github.com/gummibeer)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `mateusjunges/ignition-stackoverflow-portuguese` will be documented in this file

## 1.0.0 - 2018-09-08

- Initial release


tests/TabTest.php
<?php

namespace Junges\StackOverflowPTBR\Tests;

use Junges\StackOverflowPTBR\Tab;

class TabTest extends TestCase
{
    public function test_if_tab_has_correct_name()
    {
        $tab = new Tab();
        $this->assertEquals('StackOverflow PT-BR', $tab->name());
    }

    public function test_if_tab_has_the_correct_component()
    {
        $tab = new Tab();
        $this->assertEquals('ignition-stackoverflow-portuguese', $tab->component());
    }

    public function test_if_tab_has_the_correct_scripts()
    {
        $tab = new Tab();
        $this->assertArrayHasKey('ignition-stackoverflow-portuguese', $tab->scripts);
        $this->assertStringEndsWith('dist/js/tab.js', $tab->scripts['ignition-stackoverflow-portuguese']);
    }
}


tests/TestCase.php
<?php

namespace Junges\StackOverflowPTBR\Tests;

use Facade\Ignition\IgnitionServiceProvider;
use Orchestra\Testbench\TestCase as Orchestra;
use Junges\StackOverflowPTBR\TabServiceProvider;

abstract class TestCase extends Orchestra
{
    protected function getPackageProviders($app)
    {
        return [
            TabServiceProvider::class,
            IgnitionServiceProvider::class,
        ];
    }
}


tests/SolutionProviderTest.php
<?php

namespace Junges\StackOverflowPTBR\Tests;

use Exception;
use Throwable;
use ReflectionClass;
use Facade\IgnitionContracts\BaseSolution;
use Junges\StackOverflowPTBR\StackOverflowPTBRSolutionProvider;
use Symfony\Component\Routing\Exception\RouteNotFoundException;

class SolutionProviderTest extends TestCase
{
    public function test_if_it_can_solve_every_throwable() : void
    {
        $this->assertTrue($this->callMethod('canSolve', [
            new Exception(),
        ]));
    }

    public function test_if_it_will_call_all_tested_methods_to_get_solutions() : void
    {
        $provider = $this->getMockBuilder(StackOverflowPTBRSolutionProvider::class)
            ->onlyMethods([
                'getUrl',
                'getResponse',
                'getQuestionsByResponse',
                'getSolutionByQuestion',
            ])
            ->getMock();

        $exception = new Exception('Exception message');

        $url = 'https://api.stackexchange.com/2.2/search/advanced?page=1&pagesize=5&order=desc&sort=relevance&site=pt.stackoverflow&accepted=True&filter=!9YdnSJ*_T&q=my+exception+message';
        $questions = [
            [
                'title' => 'question_title',
                'body_markdown' => 'question_body',
                'link' => 'https://pt.stackoverflow.com',
            ],
        ];
        $response = json_encode([
            'items' => $questions,
        ]);

        $provider->expects($this->once())->method('getUrl')->with($exception)->willReturn($url);
        $provider->expects($this->once())->method('getResponse')->with($url)->willReturn($response);
        $provider->expects($this->once())->method('getQuestionsByResponse')->with($response)->willReturn($questions);
        $provider->expects($this->once())->method('getSolutionByQuestion')->with($questions[0]);

        $provider->getSolutions($exception);
    }

    public function test_it_will_return_empty_array_on_exception() : void
    {
        $provider = $this->getMockBuilder(StackOverflowPTBRSolutionProvider::class)
            ->onlyMethods([
                'getUrl',
            ])
            ->getMock();

        $exception = new Exception('Exception message');

        $provider->expects($this->once())->method('getUrl')
            ->with($exception)
            ->willReturnCallback(function (Throwable $throwable) : string {
                throw $throwable;
            });

        $this->assertEquals([], $provider->getSolutions($exception));
    }

    public function test_if_it_will_return_stackoverflow_api_url_with_exception_class() : void
    {
        $this->assertEquals(
            'https://api.stackexchange.com/2.2/search/advanced?page=1&pagesize=5&order=desc&sort=relevance&site=pt.stackoverflow&accepted=True&filter=!9YdnSJ*_T&q=Symfony%5CComponent%5CRouting%5CException%5CRouteNotFoundException',
            $this->callMethod('getUrl', [
                new RouteNotFoundException(),
            ])
        );
    }

    public function test_it_will_return_null_on_curl_error() : void
    {
        $this->assertNull($this->callMethod('getResponse', [
            'https://unexisting-domain-to-test.curl-error',
        ]));
    }

    public function test_it_will_return_null_on_timeout() : void
    {
        $this->assertNull($this->callMethod('getResponse', [
            'https://reqres.in/api/users?delay=2',
        ]));
    }

    public function it_will_return_json(): void
    {
        $this->assertJson($this->callMethod('getResponse', [
            'https://reqres.in/api/users',
        ]));
    }

    public function test_it_will_return_empty_array_if_response_is_null(): void
    {
        $this->assertEquals([], $this->callMethod('getQuestionsByResponse', [
            null,
        ]));
    }

    public function test_it_will_return_empty_array_if_response_is_invalid_json(): void
    {
        $this->assertEquals([], $this->callMethod('getQuestionsByResponse', [
            '{"this is": "invalid JSON"',
        ]));
    }

    public function test_it_will_return_empty_array_if_no_items_key_present(): void
    {
        $this->assertEquals([], $this->callMethod('getQuestionsByResponse', [
            '{"this is": "valid JSON"}',
        ]));
    }

    public function test_it_will_return_questions_array(): void
    {
        $questions = [
            [
                'title' => 'question title',
                'body_markdown' => 'question body',
                'link' => 'https://stackoverflow.com',
            ],
        ];
        $this->assertEquals($questions, $this->callMethod('getQuestionsByResponse', [
            json_encode(['items' => $questions]),
        ]));
    }

    public function test_it_will_return_null_if_title_is_missing(): void
    {
        $question = [
            'body_markdown' => 'question body',
            'link' => 'https://pt.stackoverflow.com',
        ];
        $this->assertNull($this->callMethod('getSolutionByQuestion', [
            $question,
        ]));
    }

    public function test_it_will_return_null_if_title_is_empty(): void
    {
        $question = [
            'title' => null,
            'body_markdown' => 'question body',
            'link' => 'https://pt.stackoverflow.com',
        ];
        $this->assertNull($this->callMethod('getSolutionByQuestion', [
            $question,
        ]));
    }

    public function test_it_will_return_null_if_body_is_missing(): void
    {
        $question = [
            'title' => 'question title',
            'link' => 'https://pt.stackoverflow.com',
        ];
        $this->assertNull($this->callMethod('getSolutionByQuestion', [
            $question,
        ]));
    }

    public function test_it_will_return_null_if_body_is_empty(): void
    {
        $question = [
            'title' => 'question title',
            'body_markdown' => '',
            'link' => 'https://stackoverflow.com',
        ];
        $this->assertNull($this->callMethod('getSolutionByQuestion', [
            $question,
        ]));
    }

    public function test_it_will_return_null_if_link_is_missing(): void
    {
        $question = [
            'title' => 'question title',
            'body_markdown' => 'question body',
        ];
        $this->assertNull($this->callMethod('getSolutionByQuestion', [
            $question,
        ]));
    }

    public function test_it_will_return_null_if_link_is_empty(): void
    {
        $question = [
            'title' => 'question title',
            'body_markdown' => 'question body',
            'link' => null,
        ];
        $this->assertNull($this->callMethod('getSolutionByQuestion', [
            $question,
        ]));
    }

    public function test_it_will_return_base_solution(): void
    {
        $question = [
            'title' => 'question title',
            'body_markdown' => 'question body',
            'link' => 'https://stackoverflow.com',
        ];
        /** @var BaseSolution $solution */
        $solution = $this->callMethod('getSolutionByQuestion', [
            $question,
        ]);
        $this->assertInstanceOf(BaseSolution::class, $solution);
        $this->assertEquals('question title', $solution->getSolutionTitle());
        $this->assertEquals('question body', $solution->getSolutionDescription());
        $this->assertEquals([
            'question title' => 'https://stackoverflow.com',
        ], $solution->getDocumentationLinks());
    }

    public function test_it_will_return_base_solution_with_processed_properties(): void
    {
        $question = [
            'title' => 'question &quot;title&quot;',
            'body_markdown' => 'Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.',
            'link' => 'https://pt.stackoverflow.com',
        ];

        $solution = $this->callMethod('getSolutionByQuestion', [
            $question,
        ]);

        $this->assertInstanceOf(BaseSolution::class, $solution);
        $this->assertEquals('question "title"', $solution->getSolutionTitle());
        $this->assertEquals('Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. ...', $solution->getSolutionDescription());
        $this->assertEquals([
            'question "title"' => 'https://pt.stackoverflow.com',
        ], $solution->getDocumentationLinks());
    }

    /**
     * @param string $method
     * @param array $args
     * @return mixed
     * @throws \ReflectionException
     */
    protected function callMethod(string $method, array $args)
    {
        $class = new ReflectionClass(StackOverflowPTBRSolutionProvider::class);
        $method = $class->getMethod($method);
        $method->setAccessible(true);

        return $method->invokeArgs(new StackOverflowPTBRSolutionProvider(), $args);
    }
}


src/StackOverflowSolutionProvider.php
<?php

namespace Junges\StackOverflowPTBR;

use Exception;
use Throwable;
use Illuminate\Support\Str;
use Facade\IgnitionContracts\BaseSolution;
use Facade\IgnitionContracts\HasSolutionsForThrowable;

class StackOverflowSolutionProvider implements HasSolutionsForThrowable
{
    /**
     * @param Throwable $throwable
     * @return bool
     */
    public function canSolve(Throwable $throwable): bool
    {
        return true;
    }

    /**
     * @param Throwable $throwable
     * @return array
     */
    public function getSolutions(Throwable $throwable): array
    {
        try {
            $url = $this->getUrl($throwable);
            $response = $this->getResponse($url);
            $questions = $this->getQuestionsByResponse($response);
            return array_filter(array_map([$this, 'getSolutionByQuestion'], $questions));
        } catch (Exception $exception) {
            return [];
        }
    }

    /**
     * @param array $question
     * @return BaseSolution|null
     *
     */
    protected function getSolutionByQuestion(array $question): ?BaseSolution
    {
        if (empty($question['title'])) {
            return null;
        }
        if (empty($question['body_markdown'])) {
            return null;
        }
        if (empty($question['link'])) {
            return null;
        }
        $title = html_entity_decode($question['title'], ENT_QUOTES);
        $description = Str::words(html_entity_decode($question['body_markdown'], ENT_QUOTES), 50, ' ...');
        $link = $question['link'];
        return BaseSolution::create($title)
            ->setSolutionDescription($description)
            ->setDocumentationLinks([$title => $link]);
    }

    /**
     * @param string|null $response
     * @return array
     */
    protected function getQuestionsByResponse(?string $response): array
    {
        if ($response === null) {
            return [];
        }
        $data = json_decode($response, true);
        if (json_last_error() !== JSON_ERROR_NONE) {
            return [];
        }
        return $data['items'] ?? [];
    }

    /**
     * @param string $url
     * @return string|null
     */
    protected function getResponse(string $url): ?string
    {
        $curl = curl_init($url);
        curl_setopt($curl, CURLOPT_ENCODING, 'gzip');
        curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($curl, CURLOPT_CONNECTTIMEOUT_MS, 500);
        curl_setopt($curl, CURLOPT_TIMEOUT_MS, 1000);
        $response = curl_exec($curl);
        curl_close($curl);
        if (empty($response)) {
            return null;
        }
        return $response;
    }

    /**
     * @param Throwable $throwable
     * @return string
     */
    protected function getUrl(Throwable $throwable): string
    {
        $query = $throwable->getMessage();
        if (empty($query)) {
            $query = get_class($throwable);
        }
        $query = http_build_query([
            'page' => 1,
            'pagesize' => 5,
            'order' => 'desc',
            'sort' => 'relevance',
            'site' => 'pt.stackoverflow',
            'accepted' => 'True',
            'filter' => '!9YdnSJ*_T',
            'q' => urlencode($query),
        ]);
        return 'https://api.stackexchange.com/2.2/search/advanced?'.urldecode($query);
    }
}


src/StackOverflowPTBRSolutionProvider.php
<?php

namespace Junges\StackOverflowPTBR;

use Exception;
use Throwable;
use Illuminate\Support\Str;
use Facade\IgnitionContracts\BaseSolution;
use Facade\IgnitionContracts\HasSolutionsForThrowable;

class StackOverflowPTBRSolutionProvider implements HasSolutionsForThrowable
{
    /**
     * @param Throwable $throwable
     * @return bool
     */
    public function canSolve(Throwable $throwable): bool
    {
        return true;
    }

    /**
     * @param Throwable $throwable
     * @return array
     */
    public function getSolutions(Throwable $throwable): array
    {
        try {
            $url = $this->getUrl($throwable);
            $response = $this->getResponse($url);
            $questions = $this->getQuestionsByResponse($response);

            return array_filter(array_map([$this, 'getSolutionByQuestion'], $questions));
        } catch (Exception $exception) {
            return [];
        }
    }

    /**
     * @param array $question
     * @return BaseSolution|null
     */
    protected function getSolutionByQuestion(array $question): ?BaseSolution
    {
        if (empty($question['title'])) {
            return null;
        }
        if (empty($question['body_markdown'])) {
            return null;
        }
        if (empty($question['link'])) {
            return null;
        }
        $title = html_entity_decode($question['title'], ENT_QUOTES);
        $description = Str::words(html_entity_decode($question['body_markdown'], ENT_QUOTES), 50, ' ...');
        $link = $question['link'];

        return BaseSolution::create($title)
            ->setSolutionDescription($description)
            ->setDocumentationLinks([$title => $link]);
    }

    /**
     * @param string|null $response
     * @return array
     */
    protected function getQuestionsByResponse(?string $response): array
    {
        if ($response === null) {
            return [];
        }
        $data = json_decode($response, true);
        if (json_last_error() !== JSON_ERROR_NONE) {
            return [];
        }

        return $data['items'] ?? [];
    }

    /**
     * @param string $url
     * @return string|null
     */
    protected function getResponse(string $url): ?string
    {
        $curl = curl_init($url);
        curl_setopt($curl, CURLOPT_ENCODING, 'gzip');
        curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($curl, CURLOPT_CONNECTTIMEOUT_MS, 500);
        curl_setopt($curl, CURLOPT_TIMEOUT_MS, 1000);
        $response = curl_exec($curl);
        curl_close($curl);
        if (empty($response)) {
            return null;
        }

        return $response;
    }

    /**
     * @param Throwable $throwable
     * @return string
     */
    protected function getUrl(Throwable $throwable): string
    {
        $query = $throwable->getMessage();
        if (empty($query)) {
            $query = get_class($throwable);
        }
        $query = http_build_query([
            'page' => 1,
            'pagesize' => 5,
            'order' => 'desc',
            'sort' => 'relevance',
            'site' => 'pt.stackoverflow',
            'accepted' => 'True',
            'filter' => '!9YdnSJ*_T',
            'q' => urlencode($query),
        ]);

        return 'https://api.stackexchange.com/2.2/search/advanced?'.urldecode($query);
    }
}
